extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/exec/ControllerImpl.java

```diff
@@ -0,0 +1,2040 @@
+        );
+      }
+
+      if (!finalShuffleStageDef.doesShuffle()) {
+        finalShuffleStageDef = null;
+      }
+
+      // Add all query stages.
+      // Set shuffleCheckHasMultipleValues on the stage that serves as input to the final segment-generation stage.
+      final QueryDefinitionBuilder builder = QueryDefinition.builder();
+
+      for (final StageDefinition stageDef : queryDef.getStageDefinitions()) {
+        if (stageDef.equals(finalShuffleStageDef)) {
+          builder.add(StageDefinition.builder(stageDef).shuffleCheckHasMultipleValues(true));
+        } else {
+          builder.add(StageDefinition.builder(stageDef));
+        }
+      }
+
+      // Then, add a segment-generation stage.
+      final DataSchema dataSchema = generateDataSchema(querySpec, querySignature, queryClusterBy, columnMappings);
+      builder.add(
+          StageDefinition.builder(queryDef.getNextStageNumber())
+                         .inputs(new StageInputSpec(queryDef.getFinalStageDefinition().getStageNumber()))
+                         .maxWorkerCount(tuningConfig.getMaxNumWorkers())
+                         .processorFactory(
+                             new SegmentGeneratorFrameProcessorFactory(
+                                 dataSchema,
+                                 columnMappings,
+                                 tuningConfig
+                             )
+                         )
+      );
+
+      return builder.build();
+    } else if (querySpec.getDestination() instanceof TaskReportMSQDestination) {
+      return queryDef;
+    } else {
+      throw new ISE("Unsupported destination [%s]", querySpec.getDestination());
+    }
+  }
+
+  private static DataSchema generateDataSchema(
+      MSQSpec querySpec,
+      RowSignature querySignature,
+      ClusterBy queryClusterBy,
+      ColumnMappings columnMappings
+  )
+  {
+    final DataSourceMSQDestination destination = (DataSourceMSQDestination) querySpec.getDestination();
+    final boolean isRollupQ = isRollupQuery(querySpec.getQuery());
+
+    final Pair<List<DimensionSchema>, List<AggregatorFactory>> dimensionsAndAggregators =
+        makeDimensionsAndAggregatorsForIngestion(
+            querySignature,
+            queryClusterBy,
+            destination.getSegmentSortOrder(),
+            columnMappings,
+            isRollupQ,
+            querySpec.getQuery()
+        );
+
+    return new DataSchema(
+        destination.getDataSource(),
+        new TimestampSpec(ColumnHolder.TIME_COLUMN_NAME, "millis", null),
+        new DimensionsSpec(dimensionsAndAggregators.lhs),
+        dimensionsAndAggregators.rhs.toArray(new AggregatorFactory[0]),
+        makeGranularitySpecForIngestion(isRollupQ, querySpec.getQuery()),
+        new TransformSpec(null, Collections.emptyList())
+    );
+  }
+
+  private static GranularitySpec makeGranularitySpecForIngestion(boolean isRollupQ, Query<?> query)
+  {
+    if (!isRollupQ) {
```

Comment:
Nit: since we have logic in both branches of the if, maybe reverse them and remove the negation in the if condition.

Or, since this is primarily about rollup, emit the else clause and pop the rollup logic out a level.

Reply:
I'll flip them. Agree it looks better that way.